mbedtls: some enhancement for PSA crypto core

valeriosetti opened the pull request:
- Enable legacy `CONFIG_MBEDTLS_CIPHER_AES_ENABLED` when `CONFIG_PSA_WANT_KEY_TYPE_AES` is set and Mbed TLS is the PSA crypto API provider (i.e. `CONFIG_MBEDTLS_PSA_CRYPTO_C`). This mimic what happens in Mbed TLS at build time.
- Enable `CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` instead of `CONFIG_MBEDTLS_PSA_CRYPTO_LEGACY_RNG` if the platform has any CSPRNG enabled. This helps reducing  RAM/ROM footprint.

valeriosetti committed:
mbedtls: auto-select MBEDTLS_CIPHER_AES_ENABLED when built-in in PSA

Auto-select MBEDTLS_CIPHER_AES_ENABLED when AES support is requested
through PSA (i.e. CONFIG_PSA_WANT_KEY_TYPE_AES) and the PSA support is
provided through Mbed TLS itself (i.e. CONFIG_MBEDTLS_PSA_CRYPTO_C).

This mimic what happens in Mbed TLS at build time: if AES support
is required through PSA, but there's no one else providing it
(i.e. no TF-M in Zephyr) then provide this support through legacy
AES module.

This is useful in samples/tests so that the user can simply use the
PSA_WANT symbol to ask for AES support in PSA crypto and then tune
the AES features (ex: CONFIG_MBEDTLS_AES_ROM_TABLES) without the need
to also define CONFIG_MBEDTLS_CIPHER_AES_ENABLED.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: use CSPRNG whenever possible as PSA random source

The main problem of MBEDTLS_PSA_CRYPTO_LEGACY_RNG is that it
brings in some legacy modules (entropy + ctr_drbg/hmac_drbg)
which means extra ROM/RAM footprint.
MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG instead simply calls to the
CSPRNG which makes it definitely smaller.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

dkalowsk commented:
As this has a migration guide for 4.0 change, I'm guessing this is desired to be in 4.0.  Given that we're at RC2 now, it's going to need a GitHub Issue to make that happen.  I'm 90% sure this can just be linked to the deprecation of TinyCrypt but I'll leave that up to you :-)  If you think this should be part of 4.0 please add the Milestone.

valeriosetti commented:
> As this has a migration guide for 4.0 change, I'm guessing this is desired to be in 4.0. Given that we're at RC2 now, it's going to need a GitHub Issue to make that happen. I'm 90% sure this can just be linked to the deprecation of TinyCrypt but I'll leave that up to you :-) If you think this should be part of 4.0 please add the Milestone.

This is only required for #79931 which is still drafted, so no this won't be part of 4.0. I'll move the documentation

valeriosetti committed:
mbedtls: MBEDTLS_ENTROPY_POLL_ZEPHYR default on if MBEDTLS_ENTROPY_C

As long as MBEDTLS_ENTROPY_C is enabled, Mbed TLS needs to
poll some entropy source to gather data that will then be
processed by CTR/HMAC-DRBG modules. This means that in most
of the cases, once MBEDTLS_ENTROPY_C is enabled then also
MBEDTLS_ENTROPY_POLL_ZEPHYR needs to be enabled. This was
done manually until now, as the long list of samples/tests
demonstrate.

This commit solves this dependency by defaulting
MBEDTLS_ENTROPY_POLL_ZEPHYR to on as soon as
MBEDTLS_ENTROPY_C is set. As a consequence, all manual
enablement of MBEDTLS_ENTROPY_POLL_ZEPHYR in samples/tests
are removed.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

dkalowsk commented:
As discussed in the Release meeting today, we want to consider the deprecation of Tinycrypt a release blocker item for 4.0. That would mean Part1, Part2, and the mbedTLS changes would all need to be merged in.

@ceolin any objection to this course of action? The point was raised that Tinycrypt has been "just about to be deprecated" for 5 years now, but never seems to make it. We'd like to make it happen now.

str4t0m commented on tests/crypto/mbedtls_psa/testcase.yaml:
```diff
@@ -34,8 +32,11 @@ tests:
     # Pick a platform which does not have an entropy driver. In this case we
     # enable the timer random generator because it's always available on all
     # platforms.
+    # Explicitly select CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG because this is
+    # not "automatically selected" when there is CSPRNG available.
```
Shouldn't it be: when there is **no** CSPRNG available in the last part of this comment?

valeriosetti replied:
You're absolutely right. Thanks

tomi-font commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -75,6 +75,12 @@ Mbed TLS
   be enabled. (:github:`77657`)
 
+* If a platform has a CSPRNG source available (i.e. :kconfig:option:`CONFIG_CSPRNG_ENABLED`
+  is set), then the Kconfig option :kconfig:option:`CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`
+  is the default choice for random number source instead of
```
```suggestion
  is set), the Kconfig option :kconfig:option:`CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`
  is now the default choice for the random number source instead of
```

tomi-font commented:
please review @ceolin

frkv reviewed (approved):
Good enhancements! Hope this gets in quickly

tomi-font commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -30,6 +30,12 @@ Modules
 
+* If a platform has a CSPRNG source available (i.e. :kconfig:option:`CONFIG_CSPRNG_ENABLED`
+  is set), then the Kconfig option :kconfig:option:`CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`
+  is the default choice for random number source instead of
```
```suggestion
  is now the default choice for random number source instead of
```

tomi-font reviewed (approved):
Minor nit I just realized. Thanks for the double rebase!

tomi-font commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -30,6 +30,12 @@ Modules
+  is set), then the Kconfig option :kconfig:option:`CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`
+  is the default choice for random number source instead of
+  :kconfig:option:`CONFIG_MBEDTLS_PSA_CRYPTO_LEGACY_RNG`. This helps in reducing
+  ROM/RAM footprint of the Mbed TLS library.
```
```suggestion
  the ROM/RAM footprint of the Mbed TLS library.
```

valeriosetti commented:
@ceolin now that 4.0 is out, the CI of this PR is green and we have some approval, can we please merge it?
This would reduce the dependency chain for #79931

tomi-font commented:
> @ceolin now that 4.0 is out, the CI of this PR is green and we have some approval, can we please merge it? This would reduce the dependency chain for #79931

As per the rules the assignee (so @ceolin) needs to approve the PR for it to be mergeable.
